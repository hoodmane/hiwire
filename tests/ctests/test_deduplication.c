#include "testlib.h"

int
main()
{
  __externref_t o1 = get_obj(7);
  __externref_t o2 = get_obj(9);
  HwRef id11 = hiwire_new(o1);
  ASSERT(!SLOT_DEDUPLICATED(id11));
  HwRef id12 = hiwire_new_deduplicate(o1);
  ASSERT(SLOT_DEDUPLICATED(id12));
  HwRef id13 = hiwire_new(o1);
  HwRef id14 = hiwire_new_deduplicate(o1);
  ASSERT(id11 != id12);
  ASSERT(id12 == id14);
  ASSERT(HEAP_REF_TO_INDEX(id11) == 1);
  ASSERT(HEAP_REF_TO_INDEX(id12) == 2);
  ASSERT(HEAP_REF_TO_INDEX(id13) == 3);
  ASSERT(HEAP_REF_TO_INDEX(id14) == 2);
  ASSERT(SLOT_REFCOUNT(id11) == 1);
  ASSERT(SLOT_REFCOUNT(id12) == 2);
  ASSERT(SLOT_REFCOUNT(id13) == 1);
  ASSERT(!SLOT_DEDUPLICATED(id11));
  ASSERT(SLOT_DEDUPLICATED(id12));
  ASSERT(!SLOT_DEDUPLICATED(id13));

  HwRef id21 = hiwire_new(o2);
  HwRef id22 = hiwire_new_deduplicate(o2);
  HwRef id23 = hiwire_new(o2);
  HwRef id24 = hiwire_new_deduplicate(o2);
  ASSERT(id21 != id22);
  ASSERT(id22 == id24);
  ASSERT(id22 != id12);

  hiwire_decref(id21);
  ASSERT(is_null(hiwire_get(id21)));
  hiwire_decref(id22);
  ASSERT(!is_null(hiwire_get(id22)));
  hiwire_decref(id22);
  ASSERT(is_null(hiwire_get(id22)));
  ASSERT(!is_null(hiwire_get(id23)));
  HwRef id31 = hiwire_new(o2);
  ASSERT(!SLOT_DEDUPLICATED(id31));
  HwRef id32 = hiwire_new_deduplicate(o2);
  ASSERT(SLOT_DEDUPLICATED(id32));
  ASSERT(id31 != id22);
  HwRef id33 = hiwire_new(o2);
  HwRef id34 = hiwire_new_deduplicate(o2);
  ASSERT(is_null(hiwire_get(id21)));
  ASSERT(id31 != id32);
  ASSERT(id32 == id34);
  ASSERT(id21 != id31);

  return 0;
}
