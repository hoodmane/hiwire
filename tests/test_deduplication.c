#include "testlib.h"


int main() {
    __externref_t o1 = get_obj(7);
    __externref_t o2 = get_obj(9);
    JsRef id11 = hiwire_new_value(o1);
    ASSERT(!SLOT_DEDUPLICATED(id11));
    JsRef id12 = hiwire_incref_deduplicate(id11);
    ASSERT(SLOT_DEDUPLICATED(id11));
    JsRef id13 = hiwire_new_value(o1);
    JsRef id14 = hiwire_incref_deduplicate(id13);
    ASSERT(id11 == id12);
    ASSERT(id12 == id14);
    ASSERT(HEAP_REF_TO_INDEX(id11) == 1);
    ASSERT(HEAP_REF_TO_INDEX(id13) == 2);
    ASSERT(SLOT_REFCOUNT(id11) == 3);
    ASSERT(SLOT_REFCOUNT(id13) == 1);
    ASSERT(SLOT_DEDUPLICATED(id11));
    ASSERT(!SLOT_DEDUPLICATED(id13));

    JsRef id21 = hiwire_new_value(o2);
    JsRef id22 = hiwire_incref_deduplicate(id21);
    JsRef id23 = hiwire_new_value(o2);
    JsRef id24 = hiwire_incref_deduplicate(id23);
    ASSERT(id21 == id22);
    ASSERT(id22 == id24);
    ASSERT(id22 != id12);

    hiwire_decref(id21);
    ASSERT(!is_null(hiwire_get_value(id21)));
    hiwire_decref(id21);
    ASSERT(!is_null(hiwire_get_value(id21)));
    hiwire_decref(id21);
    ASSERT(is_null(hiwire_get_value(id21)));
    ASSERT(!is_null(hiwire_get_value(id23)));
    JsRef id31 = hiwire_new_value(o2);
    ASSERT(!SLOT_DEDUPLICATED(id31));
    JsRef id32 = hiwire_incref_deduplicate(id31);
    ASSERT(SLOT_DEDUPLICATED(id31));
    JsRef id33 = hiwire_new_value(o2);
    JsRef id34 = hiwire_incref_deduplicate(id33);
    ASSERT(is_null(hiwire_get_value(id21)));
    ASSERT(id31 == id32);
    ASSERT(id32 == id34);
    ASSERT(id21 != id31);
    return 0;
}