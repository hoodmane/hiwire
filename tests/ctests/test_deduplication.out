[94;1mPASS [0m   SLOT_DEDUPLICATED(id11)
[94;1mPASS [0m   !SLOT_DEDUPLICATED(id12)
[94;1mPASS [0m   id11 != id12
[94;1mPASS [0m   id11 != id13
[94;1mPASS [0m   id12 != id13
[94;1mPASS [0m   id11 == id14
[94;1mPASS [0m   HEAP_REF_TO_INDEX(id11) == 1
[94;1mPASS [0m   HEAP_REF_TO_INDEX(id12) == 2
[94;1mPASS [0m   HEAP_REF_TO_INDEX(id13) == 3
[94;1mPASS [0m   HEAP_REF_TO_INDEX(id14) == 1
[94;1mPASS [0m   SLOT_REFCOUNT(id11) == 2
[94;1mPASS [0m   SLOT_REFCOUNT(id12) == 1
[94;1mPASS [0m   SLOT_REFCOUNT(id13) == 1
[94;1mPASS [0m   SLOT_DEDUPLICATED(id11)
[94;1mPASS [0m   !SLOT_DEDUPLICATED(id12)
[94;1mPASS [0m   !SLOT_DEDUPLICATED(id13)
[94;1mPASS [0m   id21 != id22
[94;1mPASS [0m   id22 == id24
[94;1mPASS [0m   id22 != id12
[94;1mPASS [0m   is_null(hiwire_get(id21))
[94;1mPASS [0m   !is_null(hiwire_get(id22))
[94;1mPASS [0m   is_null(hiwire_get(id22))
[94;1mPASS [0m   !is_null(hiwire_get(id23))
[94;1mPASS [0m   !SLOT_DEDUPLICATED(id31)
[94;1mPASS [0m   SLOT_DEDUPLICATED(id32)
[94;1mPASS [0m   id31 != id22
[94;1mPASS [0m   is_null(hiwire_get(id21))
[94;1mPASS [0m   id31 != id32
[94;1mPASS [0m   id32 == id34
[94;1mPASS [0m   id21 != id31
